### PR TITLE
Implement own structs and composite/sobjects request

### PR DESCRIPTION
**Pull request by rschef**
See #25, #26

Implemented:

- `Request` and `Response` structs

- `ExForce.Client` behaviour and `ExForce.Client.Tesla` adapter

- [composite/sobjects request](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobjects_collections_update.htm) to update multiple records with a single call

### CODE REVIEW

File: lib/ex_force.ex [hunk 1]
```diff
@@ -4,12 +4,14 @@ defmodule ExForce do
...
 
-  The package can be installed by adding `ex_force` to your list of dependencies in `mix.exs`:
+  The package can be installed by adding `ex_force` and `tesla` (optional if you implement your own ExForce.Client) to your list of dependencies in `mix.exs`:
 
   ```elixir
   def deps do
     [
-      {:ex_force, "~> 0.3"}
+      {:ex_force, "~> 0.3"},
+      # Optional if you implement your own ExForce.Client
```
Comment (chulkilee) [inline lib/ex_force.ex]: The current code still *require* tesla. To make it optional, we need to load the default client only when tesla is available. Let's support "bring your own client module" later - let's keep requiring tesla for now.

File: lib/ex_force/client/client.ex [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  config :ex_force, client: ClientMock
+  ```
+  """
+
+  alias ExForce.{
+    Request,
+    Response
+  }
+
+  @type client :: any()
```
Comment (chulkilee) [inline lib/ex_force/client/client.ex]: Usually `t` in a module defining a struct represents the type of the struct. So let's not add another type (`ExForce.Client.client`) here. `ExForce.Client.t` is enough.

The common pattern is, add this to the top-level module (E.g. `ExForce.client` type)

File: lib/ex_force/request.ex [hunk 3]
```diff
@@ -0,0 +1,24 @@
+defmodule ExForce.Request do
+  @moduledoc """
+  Represents an ExForce API request.
```
Comment (chulkilee) [inline lib/ex_force/request.ex]: Sounds like "ExForce API" request. Probably better to describe what it conceptually represents.

Like: 

> Represents an API request.

File: lib/ex_force/request.ex [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  """
+
+  @type param() :: binary() | [{binary() | atom(), param()}]
+  @type t :: %__MODULE__{
+          body: any(),
+          headers: [{binary(), binary()}],
+          method: :head | :get | :delete | :trace | :options | :post | :put | :patch,
+          opts: [any()],
+          query: [{binary() | atom(), param()}],
```
Comment (chulkilee) [inline lib/ex_force/request.ex]: Hm... does tesla support nested params here? If we don't use any nested params, let's just make it flat query (not including `param` inside `query`)

By nested params I mean:

```elixir
[parent: [child1: val1, child2: val2]]
```

File: lib/ex_force/response.ex [hunk 5]
```diff
@@ -0,0 +1,26 @@
+defmodule ExForce.Response do
+  @moduledoc """
+  Represents an ExForce API response.
```
Comment (chulkilee) [inline lib/ex_force/response.ex]: Same here - drop `ExForce`

File: lib/ex_force/response.ex [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  """
+
+  @type param() :: binary() | [{binary() | atom(), param()}]
+  @type t :: %__MODULE__{
+          body: any(),
+          headers: [{binary(), binary()}],
+          method: :head | :get | :delete | :trace | :options | :post | :put | :patch,
+          opts: [any()],
+          query: [{binary() | atom(), param()}],
```
Comment (chulkilee) [inline lib/ex_force/response.ex]: Same here for param in query

File: mix.exs [hunk 7]
```diff
@@ -32,14 +39,27 @@ defmodule ExForce.Mixfile do
...
+      # Lint
+      {:credo, "~> 1.1", only: [:dev, :test], runtime: false},
+      {:dialyxir, "~> 0.5", only: [:dev, :test], runtime: false},
+      # Testing
+      {:bypass, "~> 1.0", only: :test},
+      {:excoveralls, "~> 0.12", only: :test}
+    ]
+  end
+
+  defp aliases do
```
Comment (chulkilee) [inline mix.exs]: 👍  for having `test.all` 

but let's move aliases and deps reorg into separate MR so that we can merge it first!

File: lib/ex_force.ex (hunk 1 above)
Comment (rschef) [reply lib/ex_force.ex]: Fixed with 192868dd0d25388b6d23f6c77398f78e606413ce

File: lib/ex_force/client/client.ex (hunk 2 above)
Comment (rschef) [reply lib/ex_force/client/client.ex]: Fixed with dc288e04fd2dc5e95931f73bc9de9a4acf871ea1

Also moved file to lib/ex_force directory in 4e416920451217efee3ef27224287dab72708c53 in order to follow module naming

File: lib/ex_force/request.ex (hunk 3 above)
Comment (rschef) [reply lib/ex_force/request.ex]: Fixed with cdd516e792e1707d72a4cf1eb0e7961879069bdd

File: lib/ex_force/response.ex (hunk 5 above)
Comment (rschef) [reply lib/ex_force/response.ex]: Fixed with cdd516e792e1707d72a4cf1eb0e7961879069bdd

File: lib/ex_force/request.ex (hunk 4 above)
Comment (rschef) [reply lib/ex_force/request.ex]: Yes, [tesla supports nested params](https://hexdocs.pm/tesla/Tesla.Env.html#t:query/0), but I agree, let's keep it flat.

Implemented in 64ad54e108e733b617ceba9701347e181e2eacce

File: lib/ex_force/response.ex (hunk 6 above)
Comment (rschef) [reply lib/ex_force/response.ex]: Fixed with 64ad54e108e733b617ceba9701347e181e2eacce

File: mix.exs (hunk 7 above)
Comment (rschef) [reply mix.exs]: Alright 👍 

Done in ae9e5cc4794e1e63cd6fd35a2d49659cdda261b6

**Comment by chulkilee** [comment]
Merged with squash - thanks a lot! :heart:

**Comment by rschef** [comment]
> Merged with squash - thanks a lot! ❤️

Thank you, @chulkilee! Could you please publish a new release to [hex.pm](https://hex.pm/packages/ex_force)?

**Comment by chulkilee** [comment]
Released - https://hex.pm/packages/ex_force/0.4.0

I just found that I didn't include your name in the commit log - GitHub picked up the author of the first commit of this MR 🤦‍♂  I mentioned you at [CHANGELOG.md](https://github.com/chulkilee/ex_force/blob/v0.4.0/CHANGELOG.md) - sorry about that!

**Comment by rschef** [comment]
No problem, thanks! 😄
